imap/imapparse.c

```diff
@@ -259,160 +259,112 @@ EXPORTED int getxstring(struct protstream *pin, struct protstream *pout,
 {
-    uint32_t result = 0;
-    int c = EOF;
-    int gotchar = 0;
-
-    /* UINT_MAX == 4294967295U */
-    while ((c = prot_getc(pin)) != EOF && cyrus_isdigit(c)) {
-        if (result > 429496729 || (result == 429496729 && (c > '5')))
-            fatal("num too big", EX_IOERR);
-        result = result * 10 + c - '0';
-        gotchar = 1;
-    }
-
-    if (!gotchar) {
-        if (c != EOF) prot_ungetc(c, pin);
-        return EOF;
-    }
+    uint64_t result;
+    int c = _getint(pin, INT_MAX, &result);
 
     *num = result;
 
     return c;
 }
 
-EXPORTED int getint64(struct protstream *pin, int64_t *num)
+EXPORTED int getsint32(struct protstream *pin, int32_t *num)
 {
-    int64_t result = 0;
-    int c = EOF;
-    int gotchar = 0;
-
-    /* LLONG_MAX == 9223372036854775807LL */
-    while ((c = prot_getc(pin)) != EOF && cyrus_isdigit(c)) {
-        if (result > 922337203685477580LL || (result == 922337203685477580LL && (c > '7')))
-            fatal("num too big", EX_IOERR);
-        result = result * 10 + c - '0';
-        gotchar = 1;
-    }
-
-    if (!gotchar) {
-        if (c != EOF) prot_ungetc(c, pin);
-        return EOF;
-    }
+    int64_t result;
+    int c = _getsint(pin, INT_MAX, &result);
 
     *num = result;
 
     return c;
 }
 
-/* Like getint64() but explicitly signed, i.e. negative numbers
- * are accepted */
-EXPORTED int getsint64(struct protstream *pin, int64_t *num)
+EXPORTED int getuint32(struct protstream *pin, uint32_t *num)
 {
-    int c;
-    int sgn = 1;
+    uint64_t result;
+    int c = _getint(pin, UINT_MAX, &result);
 
-    c = prot_getc(pin);
-    if (c == EOF)
-        return EOF;
-
-    if (c == '-')
-        sgn = -1;
-    else if (c == '+')
-        sgn = 1;
-    else
-        prot_ungetc(c, pin);
+    *num = result;
 
-    c = getint64(pin, num);
-    if (c == EOF)
-        return EOF;
-    /* this is slightly buggy: the number LLONG_MIN == -9223372036854775808LL
-     * is a valid signed 64b int but is not accepted */
-    if (sgn < 0)
-        *num = - (*num);
     return c;
 }
 
-/* can't flag with -1 if there is no number here, so return EOF */
-EXPORTED int getuint64(struct protstream *pin, uint64_t *num)
+EXPORTED int getint64(struct protstream *pin, int64_t *num)
 {
-    uint64_t result = 0;
-    int c = EOF;
-    int gotchar = 0;
-
-    /* ULLONG_MAX == 18446744073709551615ULL */
-    while ((c = prot_getc(pin)) != EOF && cyrus_isdigit(c)) {
-        if (result > 1844674407370955161ULL || (result == 1844674407370955161ULL && (c > '5')))
-            fatal("num too big", EX_IOERR);
-        result = result * 10 + c - '0';
-        gotchar = 1;
-    }
-
-    if (!gotchar) {
-        if (c != EOF) prot_ungetc(c, pin);
-        return EOF;
-    }
+    uint64_t result;
+    int c = _getint(pin, LLONG_MAX, &result);
 
     *num = result;
 
     return c;
 }
 
+EXPORTED int getsint64(struct protstream *pin, int64_t *num)
+{
+    return _getsint(pin, LLONG_MAX, num);
```

Comment:
Since we're using `int64_t` and `uint64_t`, we should also use `INT64_MAX` and `UINT64_MAX` (rather than guessing what native types these are typedefs of).  You'll need `#include <stdint.h>`